fix docker return

mcalmer opened the pull request:
### What does this PR do?

Make it possible to use login, pull and push from module.run and get result correct

when using state.apply module.run doing docker operations retcode
is tracked to find out if the call was successful or not.

Example:
```
mgr_registries_login:
  module.run:
    - name: dockerng.login
    - registries: {{ pillar.get('docker-registries', {}).keys() }}

mgr_image_profileupdate:
  module.run:
    - name: dockerng.sls_build
    - m_name: "{{ container_name }}"
    - base: "{{ pillar.get('imagename') }}"
    - mods: packages.profileupdate
    - dryrun: True
    - kwargs:
        entrypoint: ""
    - require:
      - module: mgr_registries_login
```

The require can only work if a failed login is seen as failed (result: false).
This is the case when setting `retcode` correctly when a dictionary is returned.

### Tests written?

No

### Commits signed with GPG?

No

terminalmage reviewed (changes requested):
Please point this at the `2018.3` branch instead, `2018.3.0rc1` will soon be converted to the 2018.3.0 release and is not getting any new commits that aren't release blockers. This means that the fix will end up in 2018.3.1.

mcalmer committed:
make it possible to use login, pull and push from module.run and detect errors

when using state.apply module.run doing docker operations retcode
is tracked to find out if the call was successful or not.

mcalmer committed:
add unit test for failed login

mcalmer commented:
@terminalmage rebased to `2018.3` branch

terminalmage commented on tests/unit/modules/test_dockermod.py:
```diff
@@ -64,6 +64,26 @@ def setUp(self):
+            with patch.object(docker_mod, '_get_client', get_client_mock):
+                with patch.dict(docker_mod.__salt__, {'cmd.run_all': MagicMock(return_value=ref_out)}):
+                    ret = docker_mod.login('portus.example.com:5000')
+                    self.assertTrue('retcode' in ret)
```
Can you change this to `self.assertIn('retcode', ret)`? It tests the same condition, but the output in case of a failure will contain additional context that isn't there in a simple True/False check.

mcalmer replied:
yes, was not aware that this exists

terminalmage commented on tests/unit/modules/test_dockermod.py:
```diff
@@ -64,6 +64,26 @@ def setUp(self):
+                with patch.dict(docker_mod.__salt__, {'cmd.run_all': MagicMock(return_value=ref_out)}):
+                    ret = docker_mod.login('portus.example.com:5000')
+                    self.assertTrue('retcode' in ret)
+                    self.assertTrue(ret['retcode'] > 0)
```
Can you change this to `self.assertNotEqual(ret['retcode'], 0)`? Again, this provides additional context (i.e. the actual value of `ret['retcode']`) in the event that the condition is not met.

mcalmer replied:
yes, makes sense

terminalmage reviewed (changes requested):
Just a couple minor changes in the tests to make the failures provide additional context. Otherwise this is good to merge.

terminalmage commented:
I disagree that the return was "broken", it worked just fine. It just did not provide a retcode, which module.run can use if present to signal success or failure. If one were to define any function which didn't provide a retcode as "broken", then probably over 90% of functions in Salt are "broken".

This is still an excellent feature addition, though. :+1:

mcalmer committed:
fix checking test results

mcalmer commented:
@terminalmage do not take "broken" too serious. I had to say it to have a chance to get it in 2018.3 and not be pushed to develop :-)
